plugins/mqtt/mqtt_util.c

```diff
@@ -0,0 +1,249 @@
+    neu_json_elem_t cert      = { .name = "cert", .t = NEU_JSON_STR };
+    neu_json_elem_t key       = { .name = "key", .t = NEU_JSON_STR };
+    neu_json_elem_t keypass   = { .name = "keypass", .t = NEU_JSON_STR };
+
+    // client-id, required
+    ret = neu_parse_param(config, &error, 1, &id);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode client-id from json");
+        free(id.v.val_str);
+        return -1;
+    } else {
+        option->clientid = id.v.val_str;
+    }
+
+    // upload-topic, optional
+    ret = neu_parse_param(config, &error, 1, &upload);
+    if (0 != ret) {
+        free(upload.v.val_str);
+    } else {
+        option->upload_topic = upload.v.val_str;
+    }
+
+    // heartbeat-topic, optional
+    ret = neu_parse_param(config, &error, 1, &heartbeat);
+    if (0 != ret) {
+        free(heartbeat.v.val_str);
+    } else {
+        option->heartbeat_topic = heartbeat.v.val_str;
+    }
+
+    // format, required
+    ret = neu_parse_param(config, &error, 1, &format);
+    if (0 != ret) {
+        option->format = 0;
+    } else {
+        option->format = format.v.val_int;
+    }
+
+    // ssl, required
+    ret = neu_parse_param(config, &error, 1, &ssl);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode ssl from json");
+        return -2;
+    } else {
+        if (false == ssl.v.val_bool) {
+            option->connection = strdup("tcp://");
+        } else {
+            option->connection = strdup("ssl://");
+
+            // ca, required if ssl enabled
+            ret = neu_parse_param(config, &error, 1, &ca);
+            if (0 != ret) {
+                plog_error(plugin, "can't decode ca from json");
+                free(ca.v.val_str);
+                return -3;
+            } else {
+                option->ca = ca.v.val_str;
+            }
+
+            // cert, optional
+            ret = neu_parse_param(config, &error, 1, &cert);
+            if (0 == ret) {
+                free(cert.v.val_str);
+            } else {
+                option->cert = cert.v.val_str;
+
+                // key, required if cert enable
+                ret = neu_parse_param(config, &error, 1, &key);
+                if (0 != ret) {
+                    plog_error(plugin, "can't decode key from json");
+                    free(key.v.val_str);
+                    return -4;
+                } else {
+                    option->key = key.v.val_str;
+                }
+
+                // keypass, optional
+                ret = neu_parse_param(config, &error, 1, &keypass);
+                if (0 != ret) {
+                    free(keypass.v.val_str);
+                } else {
+                    option->keypass = keypass.v.val_str;
+                }
+            }
+        }
+    }
+
+    // host, required
+    ret = neu_parse_param(config, &error, 1, &host);
+    if (0 != ret) {
+        plog_error(plugin, "can't decode host from json");
+        free(host.v.val_str);
+        return -5;
```

Comment:
should return NEU_ERR_NODE_SETTING_INVALID

Reply:
Return NEU_ERR_NODE_SETTING_INVALID in the outer function uniformly